devito/arguments.py

```diff
@@ -158,198 +139,381 @@ class PtrArgument(Argument):
+                                    for x in dimension_parameters])
+
+        other_arguments = [ArgumentVisitor().visit(x) for x in parameters
+                           if x not in tensor_arguments + scalar_arguments]
+
+        return tensor_arguments + dimension_parameters + scalar_arguments +\
+            other_arguments
+
+    def _extract_children_of_composites(self, kwargs):
+        new_params = {}
+        # If we've been passed CompositeFunction objects as kwargs,
+        # they might have children that need to be substituted as well.
+        for k, v in kwargs.items():
+            if isinstance(v, CompositeFunction):
+                orig_param_l = [i for i in self.parameters if i.name == k]
+                # If I have been passed a parameter, I must have seen it before
+                if len(orig_param_l) == 0:
+                    raise InvalidArgument("Parameter %s does not exist in expressions " +
+                                          "passed to this Operator" % k)
+                # We've made sure the list isn't empty. Names should be unique so it
+                # should have exactly one entry
+                assert(len(orig_param_l) == 1)
+                orig_param = orig_param_l[0]
+                # Pull out the children and add them to kwargs
+                for orig_child, new_child in zip(orig_param.children, v.children):
+                    new_params[orig_child.name] = new_child
+        kwargs.update(new_params)
+        return kwargs
+
+    def _dle_arguments(self, dim_sizes):
+        # Add user-provided block sizes, if any
+        dle_arguments = OrderedDict()
+        autotune = True
+        for i in self.dle_arguments:
+            dim_size = dim_sizes.get(i.original_dim.name, None)
+            if dim_size is None:
+                raise InvalidArgument('Unable to derive size of dimension %s from '
+                                      'defaults. Please provide an explicit '
+                                      'value.' % i.original_dim.name)
+            if i.value:
+                try:
+                    dle_arguments[i.argument.name] = i.value(dim_size)
+                except TypeError:
+                    dle_arguments[i.argument.name] = i.value
+                    autotune = False
+            else:
+                dle_arguments[i.argument.name] = dim_size
+        return dle_arguments, autotune
+
+    def _derive_values(self, kwargs):
+        """ Populate values for all the arguments. The values provided in kwargs will
+            be used wherever provided. The values for the rest of the arguments will be
+            derived from the ones provided. The default values for the tensors come from
+            the data property of the symbols used in the Operator.
         """
-        raise NotImplemented()
+        values = OrderedDict()
+
+        for i in self.arguments:
+            values[i] = get_value(i, kwargs.pop(i.name, None), values)
```

Comment:
not pythonic

Reply:
Please elaborate

Reply:
``values = OrderedDict([(i, get_value(i, kwargs.pop(i.name, None), values)) for i in self.arguments])``

Reply:
Done